### PR TITLE
Improve readability perfect minimal hash algorithm

### PR DESCRIPTION
**Pull request by jaydee-io**
Refactor `make_array_of_items` to improve readability

- Introduce a new `constexpr vector` type to ease manipulating arrays.
- Change `partition` and `quicksort` algorithms to take iterators instead of array and indices
- Remove unused `sort` algorithm

### CODE REVIEW

**Comment by serge-sans-paille** [comment]
WoW, nice refactoring. I'll chec this in details, but at first glance that looks very nice to me.

File: include/frozen/bits/algorithms.h [hunk 1]
```diff
@@ -87,63 +87,35 @@ constexpr void cswap(std::tuple<Tys...> &a, std::tuple<Tys...> &b) {
...
 }
 
 template <typename T, std::size_t N, class Compare>
 constexpr std::array<T, N> quicksort(std::array<T, N> const &array,
                                      Compare const &compare) {
   std::array<T, N> res = array;
-  quicksort(res, 0, N - 1, compare);
+  quicksort(&std::get<0>(res), &std::get<N - 1>(res), compare);
   return res;
 }
```
Comment (serge-sans-paille) [inline include/frozen/bits/algorithms.h]: That's much cleaner with the iterator version!

File: include/frozen/bits/basic_types.h [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+#ifndef FROZEN_LETITGO_BASIC_TYPES_H
+#define FROZEN_LETITGO_BASIC_TYPES_H
+
+#include <array>
+
+namespace frozen {
+
+namespace bits {
+
+template <class T, size_t N>
```
Comment (serge-sans-paille) [inline include/frozen/bits/basic_types.h]: s/size_t /std::size_t/ there

File: include/frozen/bits/basic_types.h [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+  using const_pointer = const value_type *;
+  using iterator = pointer;
+  using const_iterator = const_pointer;
+  using size_type = std::size_t;
+  using difference_type = std::ptrdiff_t;
+
+  // Constructors
+  constexpr cvector() : data{{}} { }
+  // Only used for zero initialized vector, size is not incremented
+  constexpr cvector(T il) : data{il} { }
```
Comment (serge-sans-paille) [inline include/frozen/bits/basic_types.h]: dsize should be set to 1 here if I get it corectly

File: include/frozen/bits/basic_types.h [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+  // Iterators
+  constexpr iterator begin() noexcept { return data; }
+  constexpr iterator end() noexcept { return data + N; }
```
Comment (serge-sans-paille) [inline include/frozen/bits/basic_types.h]: N should be dsize, right?

File: include/frozen/bits/basic_types.h [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+  // Capacity
+  constexpr size_type size() const { return dsize; }
+
+  // Element access
+  constexpr       reference operator[](std::size_t index) { return data[index]; }
+  constexpr const_reference operator[](std::size_t index) const { return data[index]; }
+
+  constexpr       reference back() { return data[dsize - 1]; }
```
Comment (serge-sans-paille) [inline include/frozen/bits/basic_types.h]: incorrect if dsize == 0

File: include/frozen/bits/basic_types.h [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  constexpr const_reference back() const { return data[dsize - 1]; }
+
+  // Modifiers
+  constexpr void push_back(const T & a) { data[dsize++] = a; }
+  constexpr void push_back(T && a) { data[dsize++] = a; }
```
Comment (serge-sans-paille) [inline include/frozen/bits/basic_types.h]: missing a move here

File: include/frozen/bits/basic_types.h [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  constexpr void pop_back() { --dsize; }
+
+  constexpr void clear() { dsize = 0; }
+
+  constexpr std::array<T, N> to_array() {
```
Comment (serge-sans-paille) [inline include/frozen/bits/basic_types.h]: do you really want that? you're loosing the size info in the process...

File: include/frozen/bits/pmh.h [hunk 8]
```diff
@@ -33,23 +34,32 @@ namespace frozen {
...
+                            bucket<N> const &b1) const {
+    return b0.size() > b1.size();
   }
 };
 
+template<class T, size_t N>
+constexpr void memset(cvector<T, N> & data, T && a) {
+  for (std::size_t i = 0; i < data.size(); ++i)
+    data[i] = a;
+}
```
Comment (serge-sans-paille) [inline include/frozen/bits/pmh.h]: you could overlaod the ``operator=`` of ``cvector`` to achieve this

**Comment by jaydee-io** [comment]
In fact, I hesitated between an `array` and a `vector`.
Sometimes, we need to access the whole array (from 0 to N-1).
Sometimes, we only need to access values that have been pushed (from 0 to dsize - 1).
Maybe it would be worth splitting it into separate array and vector.

What do you think ?

File: include/frozen/bits/basic_types.h (hunk 2 above)
Comment (jaydee-io) [inline include/frozen/bits/basic_types.h]: Fixed

File: include/frozen/bits/basic_types.h (hunk 5 above)
Comment (jaydee-io) [inline include/frozen/bits/basic_types.h]: Yes, but it's the same as `std::vector::back()`:
> Calling this function on an empty container causes undefined behavior.

File: include/frozen/bits/basic_types.h (hunk 7 above)
Comment (serge-sans-paille) [inline include/frozen/bits/basic_types.h]: ok you need it to go back to regular ``std::array``

**Comment by serge-sans-paille** [comment]
@jaydee-io I'm ok with the concept of ``cvector``. there's just a few remarks that needs to be patched and it's in!

**Comment by jaydee-io** [comment]
Here we are !

**Comment by serge-sans-paille** [comment]
thanks !

**Pull request merged**